lib/spack/spack/extensions.py

```diff
@@ -11,10 +11,48 @@
 import types
 
 import llnl.util.lang
-import llnl.util.tty as tty
+import spack.cmd
 import spack.config
+from spack.error import SpackError
 
-extension_regexp = re.compile(r'spack-([\w]*)')
+_extension_regexp = re.compile(r'spack-([\w]*)')
+_extension_command_map = None
```

Comment:
Is this just an optimization, to cache names and command paths? Or is it solving a more fundamental issue?

Reply:
It's more-or-less an optimization: in the previous iteration of the code IIRC, `main()` did a sequential filesystem search through every configured extension area in order for the requested command, which seemed clunky in the case of multiple configured extensions with possibly multiple commands in each, and especially in the case of the extended help command.